### ISSUE TITLE
querier2 cannot query data after persist

### ISSUE THREAD

**Issue by mkmik** [influxdata/influxdb_iox#6421]
__Steps to reproduce and actual behavior:__

spin up `ingester2`, `router2` and `querier` after applying a small diff to work around #6420:

```diff
--- ingester2/src/init.rs
+++ ingester2/src/init.rs
@@ -173,7 +173,7 @@ pub async fn new(
         .create_or_get(&topic, TRANSITION_SHARD_INDEX)
         .await
         .expect("create transition shard");
-    assert_eq!(s.id, TRANSITION_SHARD_ID);
+    //    assert_eq!(s.id, TRANSITION_SHARD_ID);
     txn.commit().await.expect("commit transition shard");
 
     // Initialise the deferred namespace name resolver.
```

Set-up empty catalog db:

```bash
mkdir -p /tmp/iox/{wal,obj}

createdb iox_shared
./target/debug/influxdb_iox catalog setup --catalog-dsn postgres:///iox_shared

# there has to exist one "topic", see https://github.com/influxdata/influxdb_iox/issues/6420
psql 'dbname=iox_shared options=-csearch_path=public,iox_catalog' -c "insert into topic (name) values ('iox-shared')"
```

Run ingester2:

```bash
INFLUXDB_IOX_RPC_MODE=2 ./target/debug/influxdb_iox run ingester2 --api-bind=127.0.0.1:8081 --grpc-bind=127.0.0.1:8042 --wal-directory /tmp/iox/wal  --catalog-dsn postgres:///iox_shared --object-store=file --data-dir=/tmp/iox/obj -v 
```

Run router2:

```bash
INFLUXDB_IOX_RPC_MODE=2 ./target/debug/influxdb_iox run router2 --api-bind=127.0.0.1:8080 --grpc-bind=127.0.0.1:8085 --ingester-addresses=127.0.0.1:8042 --catalog-dsn postgres:///iox_shared -v
```

Run querier:

```bash
INFLUXDB_IOX_RPC_MODE=2 ./target/debug/influxdb_iox run querier --ingester-addresses=http://127.0.0.1:8042 --api-bind 127.0.0.1:8083 --grpc-bind 127.0.0.1:8082 --catalog-dsn postgres:///iox_shared --object-store=file --data-dir=/tmp/iox/obj -v
```

Load some data:

```bash
curl -i -X POST "http://localhost:8080/api/v2/write?bucket=test&org=internal&precision=s" --data-binary @./test_fixtures/lineproto/air_and_water.lp
```

Run a query. I can see the querier return the data:

```console
$ ./target/debug/influxdb_iox query internal_test 'select * from air_temperature'
+--------------+-------------------+-------+------------------------+----------------------+
| location     | sea_level_degrees | state | tenk_feet_feet_degrees | time                 |
+--------------+-------------------+-------+------------------------+----------------------+
| coyote_creek | 77.1              | CA    | 41                     | 2020-09-22T06:29:20Z |
| puget_sound  | 78                | WA    | 40.9                   | 2020-09-22T06:29:20Z |
| santa_monica | 77.6              | CA    | 40.9                   | 2020-09-22T06:29:20Z |
| coyote_creek | 77.2              | CA    | 40.8                   | 2019-09-17T21:36:00Z |
| puget_sound  | 77.5              | WA    | 41.1                   | 2019-09-17T21:36:00Z |
| santa_monica | 77.3              | CA    | 40                     | 2019-09-17T21:36:00Z |
+--------------+-------------------+-------+------------------------+----------------------+
```

The data is in the wal and not yet persisted in the object store:

```console
$ tree /tmp/iox
/tmp/iox
├── obj
└── wal
    └── 5.dat
```

waiting 5m for the data to be persisted:

```console
$ sleep 5m; tree /tmp/iox
/tmp/iox
├── obj
│   └── 1
│       ├── 1
│       │   └── 1
│       │       ├── 3
│       │       │   └── 48b98a37-9367-4f8b-9632-e2e440781763.parquet
│       │       └── 4
│       │           └── 81ea6662-17f4-487f-89b5-8854dd7ae01a.parquet
│       └── 2
│           └── 1
│               ├── 1
│               │   └── 188bfd0e-cd26-49b6-8cbf-06449da92ccf.parquet
│               └── 2
│                   └── 01e1f4f4-4f98-45b4-b991-be140d8ada41.parquet
└── wal
    └── 6.dat
```

However, I can no longer query the data:

```console
$ ./target/debug/influxdb_iox query internal_test 'select * from air_temperature'
++
++
```

I tried the query again multiple times during the 10m it took to file this ticket. 

After 10m I *restart the querier* and I can query the data:

```
$ ./target/debug/influxdb_iox query internal_test 'select * from air_temperature'
+--------------+-------------------+-------+------------------------+----------------------+
| location     | sea_level_degrees | state | tenk_feet_feet_degrees | time                 |
+--------------+-------------------+-------+------------------------+----------------------+
| coyote_creek | 77.1              | CA    | 41                     | 2020-09-22T06:29:20Z |
| santa_monica | 77.6              | CA    | 40.9                   | 2020-09-22T06:29:20Z |
| puget_sound  | 78                | WA    | 40.9                   | 2020-09-22T06:29:20Z |
| coyote_creek | 77.2              | CA    | 40.8                   | 2019-09-17T21:36:00Z |
| santa_monica | 77.3              | CA    | 40                     | 2019-09-17T21:36:00Z |
| puget_sound  | 77.5              | WA    | 41.1                   | 2019-09-17T21:36:00Z |
+--------------+-------------------+-------+------------------------+----------------------+
```

__Expected behaviour:__

I expect the querier to be able to query the data after it has been persisted.
IIUC the querier has a catalog cache, so it's not hitting the catalog to find parquet files on every query.
However I expect the ingester to tell the querier it has persisted the data so that the querier becomes aware of the new parquet files.

__Environment info:__

* Please provide the command you used to build the project, including any `RUSTFLAGS`.
* System info: Run `uname -srm` or similar and copy the output here (we want to know your OS, architecture etc).
* If you're running IOx in a containerised environment then details about that would be helpful.
* Other relevant environment details: disk info, hardware setup etc.

__Config:__
Copy any non-default config values here or attach the full config as a gist or file.

__Logs:__

querier log:

```
2022-12-16T06:13:31.988321Z  INFO service_grpc_flight: Running SQL via flight do_get namespace_name=internal_test query=select * from air_temperature trace=
```

the ingester indeed said it persisted the partitions:

```
2022-12-16T05:58:50.778246Z  INFO ingester2::wal::rotate_task: rotating wal file
2022-12-16T05:58:55.866958Z  INFO ingester2::persist::context: persisted partition object_store_id=188bfd0e-cd26-49b6-8cbf-06449da92ccf namespace_id=1 namespace_name=internal_test table_id=2 table_name=air_temperature partition_id=1 partition_key=2019-09-17 total_persist_duration=62.751041ms active_persist_duration=62.546666ms queued_persist_duration=204.375µs
2022-12-16T05:58:55.870667Z  INFO ingester2::persist::context: persisted partition object_store_id=48b98a37-9367-4f8b-9632-e2e440781763 namespace_id=1 namespace_name=internal_test table_id=1 table_name=h2o_temperature partition_id=3 partition_key=2019-09-17 total_persist_duration=66.335041ms active_persist_duration=57.004458ms queued_persist_duration=9.330583ms
2022-12-16T05:58:55.872630Z  INFO ingester2::persist::context: persisted partition object_store_id=01e1f4f4-4f98-45b4-b991-be140d8ada41 namespace_id=1 namespace_name=internal_test table_id=2 table_name=air_temperature partition_id=2 partition_key=2020-09-22 total_persist_duration=68.354916ms active_persist_duration=65.162583ms queued_persist_duration=3.192333ms
2022-12-16T05:58:55.873579Z  INFO ingester2::persist::context: persisted partition object_store_id=81ea6662-17f4-487f-89b5-8854dd7ae01a namespace_id=1 namespace_name=internal_test table_id=1 table_name=h2o_temperature partition_id=4 partition_key=2020-09-22 total_persist_duration=69.241666ms active_persist_duration=68.57375ms queued_persist_duration=667.916µs
2022-12-16T05:58:55.874264Z  INFO ingester2::wal::rotate_task: dropped persisted wal segment closed_id=5
```

**Comment by mkmik** [comment]
Ah ok, Paul told me to deploy if in staging so I assumed it was ready and didn't bother to read all the fine print.

Is there already an issue for the missing steps or shall we keep this issue open to track this?

**Comment by crepererum** [comment]
I don't recall any open issue, but @carols10cents may know more.

**Comment by carols10cents** [comment]
Invalidating the cache was done in https://github.com/influxdata/influxdb_iox/pull/6365 but may not be working correctly, I'm on it.

**Comment by carols10cents** [comment]
> ```
> # there has to exist one "topic", see https://github.com/influxdata/influxdb_iox/issues/6420
> psql 'dbname=iox_shared options=-csearch_path=public,iox_catalog' -c "insert into topic (name) values ('iox-shared')"
> ```

This can also be done with:

```
./target/debug/influxdb_iox catalog topic update iox-shared
```

**Comment by domodwyer** [comment]
This is likely because of the failing assert that was removed - the catalog is not in the correct state w.r.t the systems invariants.

**Comment by mkmik** [comment]
FWIW, the shard table of the instance where I ran this test is:

```console
$ psql 'dbname=iox_shared options=-csearch_path=public,iox_catalog' -c 'select * from shard'
 id | topic_id | shard_index | min_unpersisted_sequence_number 
----+----------+-------------+---------------------------------
  1 |        2 |           1 |                               0
```

**Comment by domodwyer** [comment]
I think the topic ID might be the cause? It should be `1` I think - I'm not sure - this is a guess, rather than a "it's that!".

**Comment by carols10cents** [comment]
I've tried taking out the TRANSITION_SHARD_ID const and getting it from the catalog for the TRANSITION_SHARD_INDEX instead; that didn't fix the problem.

What I have found is when the ingester doesn't have any record batches in memory for a query, it's not even streaming back the metadata with the persisted file count. Looking into why.

**Comment by carols10cents** [comment]
Ok, when I run one ingester2, one router2, and one querier2 locally, then run the data generator such that it is writing data every 10 seconds, and then I query for that data before and after the ingester persisting, I don't see any loss of data (the amount of data returned from the query is always growing with the newly written data and there's a query result with a timestamp every 10 seconds, no gaps).

Next I'm going to try running 2 ingester2s, 1 router2 talking to both ingester2s, and 1 querier2 talking to both ingester2s.

**Comment by carols10cents** [comment]
Ok well when I try running 2 ingester2s, 1 router2 talking to both ingester2s, and 1 querier2 talking to both ingester2s, with a data generator writing a bit of data every 10 seconds, I get this error from the querier:

```
Error querying: Tonic(Status { code: Internal, message: "Error while planning query: External error: Ingester 'http://127.0.0.1:8042' and 'http://127.0.0.1:8072' both provide data for partition 3", metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Wed, 21 Dec 2022 16:04:21 GMT", "content-length": "0"} }, source: None })
```

It's.... expceted... that both ingesters have data for the same partition, I think? Not sure why this is an error; investigating now.

**Comment by domodwyer** [comment]
Previously data for one partition was never on more than one ingester - I'm guessing there's a baked-in assumption that partitions can only exist on one ingester.

Sounds like a good candidate for causing query problems!

**Comment by carols10cents** [comment]
Ok fixed that by disabling the check in RPC write mode, PR incoming soon... next problem is that one of my 2 ingester 2s is looping with these relevant-looking logs:

```
2022-12-21T16:26:45.485633Z DEBUG ingester2::persist::context: updating partition sort key object_store_id=3769594e-d741-4bc9-a15f-fd69336b32d1 namespace_id=4 namespace_name=mlb_redsox table_id=4 table_name=traces partition_id=5 partition_key=2022-12-21 new_sort_key=SortKey { columns: {"host": SortOptions { descending: false, nulls_first: true }, "service": SortOptions { descending: false, nulls_first: true }, "trace_id": SortOptions { descending: false, nulls_first: true }, "span_id": SortOptions { descending: false, nulls_first: true }, "time": SortOptions { descending: false, nulls_first: true }} } old_sort_key=None
2022-12-21T16:26:45.486391Z  WARN ingester2::persist::context: detected concurrent sort key update, regenerating parquet object_store_id=3769594e-d741-4bc9-a15f-fd69336b32d1 namespace_id=4 namespace_name=mlb_redsox table_id=4 table_name=traces partition_id=5 partition_key=2022-12-21 expected=None observed=["host", "service", "span_id", "trace_id", "time"] update=["host", "service", "trace_id", "span_id", "time"]
```

Important parts extracted:

```
2022-12-21T16:26:45.485633Z DEBUG ingester2::persist::context: updating partition sort key 
    new_sort_key=SortKey { columns: {
        "host": SortOptions { descending: false, nulls_first: true }, 
        "service": SortOptions { descending: false, nulls_first: true }, 
        "trace_id": SortOptions { descending: false, nulls_first: true }, 
        "span_id": SortOptions { descending: false, nulls_first: true }, 
        "time": SortOptions { descending: false, nulls_first: true }} } 
    old_sort_key=None
2022-12-21T16:26:45.486391Z  WARN ingester2::persist::context: detected concurrent sort key update, regenerating parquet 
    expected=None 
    observed=["host", "service", "span_id", "trace_id", "time"] 
    update=["host", "service", "trace_id", "span_id", "time"]
```

I think it's never updating its `expected` for the compare-and-swap? I know this just changed yesterday, looking...

**Comment by domodwyer** [comment]
So long as you're running with https://github.com/influxdata/influxdb_iox/pull/6448 this should be fine to ignore (it just requires retrying the persist op).

**Comment by carols10cents** [comment]
I am running with that, and it is retrying, the problem is it appears to be retrying forever and never succeeding....

**Comment by carols10cents** [comment]
Ok Dom fixed that problem in https://github.com/influxdata/influxdb_iox/pull/6456 🎉 

I'm not seeing dropped data locally. Next I'm going to try running a compactor2 as well as 2 ingester2s, a router2, and a querier2.

If that doesn't reproduce the issue, I'm going to go look at the setup in k8s-idpe and try reproducing that more closely... if *that* doesn't work, I'll need help from someone who *is* still seeing this problem to create updated reproduction steps.

**Comment by carols10cents** [comment]
Yeah, I'm at a dead end. Adding a compactor2 in to my env didn't affect the query results I was seeing. I don't see anything in the k8s-idpe setup that seems like it would be relevant...

So once the fix for #6456 is deployed, I'd be interested to know if the results disappearing is still happening, and if so, thoughts on what I might be missing for reproducing it locally.

**Comment by domodwyer** [comment]
I suspect https://github.com/influxdata/influxdb_iox/pull/6455 has fixed this (certainly seems like it works for @carols10cents!)

@mkmik I'm going to close this based on Carol's validation work above - if it's still a problem please reopen :+1:
